packages/jsii-pacmak/lib/targets/python/type-name.ts

```diff
@@ -0,0 +1,354 @@
+  isNamedTypeReference,
+  isPrimitiveTypeReference, PrimitiveType, PrimitiveTypeReference,
+  isUnionTypeReference,
+  Type,
+} from '@jsii/spec';
+import { createHash } from 'crypto';
+import { die, toPythonIdentifier } from './util';
+
+export interface TypeName {
+  pythonType(context: NamingContext): string;
+  requiredImports(context: NamingContext): PythonImports;
+}
+
+export interface PythonImports {
+  /**
+   * For a given source module, what elements to import. The empty string value
+   * indicates a need to import the module fully ("import <name>") instead of
+   * doing a piecemeal import ("from <name> import <item>").
+   */
+  readonly [sourcePackage: string]: ReadonlySet<string>;
+}
+
+/**
+ * The context in which a PythonType is being considered.
+ */
+export interface NamingContext {
+  /** The assembly in which the PythonType is expressed. */
+  readonly assembly: Assembly;
+
+  /** The submodule of the assembly in which the PythonType is expressed (could be the module root) */
+  readonly submodule: string;
+
+  /**
+   * The declaration is made in the context of a type annotation (so it can be quoted)
+   *
+   * @default true
+   */
+  readonly typeAnnotation?: boolean;
+
+  /**
+   * The nesting scope in which the PythonType is expressed (if any)
+   *
+   * @default - none
+   */
+  readonly nestingScope?: string | undefined;
+
+  /**
+   * Disables generating typing.Optional wrappers
+   * @default false
+   * @internal
+   */
+  readonly ignoreOptional?: boolean;
+}
+
+export function toTypeName(ref?: OptionalValue | TypeReference): TypeName {
+  if (ref == null) {
+    return Primitive.NONE;
+  }
+
+  const type = isOptionalValue(ref) ? ref.type : ref;
+  const optional = isOptionalValue(ref) && ref.optional;
+
+  let result: TypeName = Primitive.ANY;
+
+  if (isPrimitiveTypeReference(type)) {
+    result = Primitive.of(type);
+  } else if (isCollectionTypeReference(type)) {
+    const elt = toTypeName(type.collection.elementtype);
+    if (type.collection.kind === CollectionKind.Array) {
+      result = new List(elt);
+    } else {
+      result = new Dict(elt);
+    }
+  } else if (isUnionTypeReference(type)) {
+    result = new Union(type.union.types.map(toTypeName));
+  } else if (isNamedTypeReference(type)) {
+    result = new UserType(type.fqn);
+  }
+
+  return optional ? new Optional(result) : result;
+}
+
+export function mergePythonImports(...pythonImports: readonly PythonImports[]): PythonImports {
+  const result: Record<string, Set<string>> = {};
+  for (const bag of pythonImports) {
+    for (const [packageName, items] of Object.entries(bag)) {
+      if (!(packageName in result)) {
+        result[packageName] = new Set();
+      }
+      for (const item of items) {
+        result[packageName].add(item);
+      }
+    }
+  }
+  return result;
+}
+
+function isOptionalValue(type: OptionalValue | TypeReference): type is OptionalValue {
+  return (type as unknown as OptionalValue).type != null;
```

Comment:
the fact that `type` has `type` indicates that it's an optional value? Sounds odd.

Reply:
`OptionalValue` is an interface that has `{ type: TypeReference, optional?: boolean }`. The only "always there" field is `type`.

Reply:
Rename `type` to `optionalType` or something like that so it will be easier to distinguish (and avoid bugs).